tests/components/modbus/test_modbus.py

```diff
@@ -50,23 +103,397 @@
             CONF_PARITY: "E",
             CONF_STOPBITS: 1,
         },
+        {
+            CONF_TYPE: "serial",
+            CONF_BAUDRATE: 9600,
+            CONF_BYTESIZE: 8,
+            CONF_METHOD: "rtu",
+            CONF_PORT: "usb01",
+            CONF_PARITY: "E",
+            CONF_STOPBITS: 1,
+            CONF_NAME: "modbusTest",
+            CONF_TIMEOUT: 30,
+            CONF_DELAY: 10,
+        },
     ],
 )
-async def test_config_modbus(hass, do_discovery, do_options, do_config):
+async def test_config_modbus(hass, do_config):
     """Run test for modbus."""
+
+    await _config_helper(hass, do_config)
+    assert DOMAIN in hass.data
+    assert len(hass.data[DOMAIN]) == 1
```

Comment:
It's a bit annoying that we do these assertions of hass.data which is an integration detail. It would be better if we could assert something in the core state or calls to a patched library, but I'm not sure if there's something.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

Reply:
We have 2 different views of "public". ModbusHub is used by all modbus platforms and in that sense "public". Without having tested it, I think it could be imported in other integrations as well, however I would not recommend that, but instead use "dependencies" in manifest.

I wanted to isolate all pymodbus (the external library) testing in test_modbus.py, so that the platforms do not need to consider how pymodbus is implemented. While doing that it felt right to test the basic modbus configuration, so that test_modbus.py tests modbus.py with pymodbus.

In order to avoid using hass.data I have been trying DeviceEntry and RegistryEntry both without success. I have earlier tried other ways without getting anywhere.

I have instead added that to my task list, because proving such a function both for test and production (you very often see platforms use hass.data[DOMAIN] to get access to the integration). That is a nice task for me to get started on core.

I hope you can live with me explanations, otherwise please give me a more detailed hint (writing test for integrations is really high level explanation).

Reply:
The tests of integrations should not access hass.data at all really. Data that is stored there is an integration detail that the tests should not check.

The common aim when checking hass.data in tests is to check that the integration has cleaned up after itself and don't store things there anymore, eg on config entry unload. It could be ok to check that nothing is stored for the integration domain in hass.data but any further checks will go into details that we should avoid.

Reply:
As long as the ModbusHub is just an implementation detail in this integration we should not test it separately but just let the platform tests check that the correct state is achieved when setting up the integration and its platform.

Reply:
Ok I found a way to avoid accessing hass.data, which is not as nice as I wanted and still do not solve the base problem “how can a platform or test get hold of the integration object, without accessing hass.data”,

_config_helper is NOT testing ModbusHub, but using async_setup_component to secure the integration can load. It is not logical to have all platforms test that the integration can load multiple times see test_config_multiple_modbus() nor is it logical that each platform test the parameters of the integration see test_config_modbus(). I can of course split the integration testing (those 2 functions) into a seperate file if you want, but not having these tests are important.

Reply:
Using `async_setup_component` is good. Our standard is to have a module `test_init.py` where we test that we can set up the integration in general and handle the config ok.

The platform tests should set up the integration with `async_setup_component` and check that the expected entity state is achieved via hass.states and that services can be called via hass.services, according to the dev docs link.

Reply:
Sorry for questioning your responses, but I want to learn, so I step by step I get a better reviewer and my PR´s are more perfect.

I have moved the integration setup tests to test_init.py so that test_modbus.py concentrates on modbusHub. There are one small part in a grey zone, the write calls. I prefer to have them in modbus.py because the first tests are purely modbusHub and the next tests are using hass.service to test the whole route.

Reply:
I'd recommend removing the tests that access ModbusHub directly and test that code indirectly in the platforms instead. All code paths will be tested anyway if we add enough tests and we'll only assert what matters, which is the library interface interaction and the core state data. Then we'll able to refactor ModbusHub if needed without changing the tests.

Reply:
My idea was slightly different, I wanted to ensure the whole class modbusHub was tested together with pymodbus, because then all following tests could mock the modbusHub class. That will allow us to refactor ModbusHub as long as the interface do not change. 

But this is a matter of style, and not important, so I have followed your recommendation for the write calls, because when making the service calls we pass the same code. I have kept the create/close/connect tests, because they are not really tested with the platform tests. I have kept the read calls because the help me secure against interface changes when I update pymodbus (but I can remove them, and keep them private if they bother you).

Reply:
My suggestion is that we don't access the ModbusHub directly at all in the tests. All tests should start by setting up the integration and then we check that the integration works via the core interfaces.

If the create/close/connect is done centrally we can test this in the test_init.py module by setting up the integration.

If we want to guard for changes in the library we can autospec the mock when patching the library.

https://docs.python.org/3/library/unittest.mock.html#autospeccing

Reply:
I have removed all direct testing, I will later add more of the testing you suggest.